### ISSUE TITLE
Grant proposal from Graficadirecta: A utility to optimize orchestrator stake vs fees for the livepeer network.

### ISSUE THREAD

**Issue by pabloVoorvaart** [livepeer/Grant-Program#25]
**Give a 3 sentence description about this proposal.**
A utility to optimize orchestrator stake vs fees for the livepeer network.

**Describe the problem you are solving.** 

I was recently interested in setting up an orchestrator / transcoder node on the livepeer network. While doing research on the possible returns on investment for the node, I found it difficult to estimate the required stake necessary to at least cover the cost of maintenance, since to my understanding, stake is the main driving force behind receiving signals from broadcasters.

While researching the forums I found streamflow transcoding financial model, which was a great starting point to understand the economics behind livepeers orchestrators. The model though had outdated data and was doing calculations for transcoding assuming an incoming traffic into the network 10x greater that the one the network currently has; around 500K minutes weekly. (Data taken from the livepeer dashboard).

After looking at streamflows model I decided to try and create a small web based utility that based on the data supplied by the livepeer thegraph api and the livepeer dashboard allowed:

1.  Orchestrators to better understand their current expected rewards / fees based on the current network traffic, and orchestrator stake. 
2.  Orchestrators to better understand how much delegated stake they need to reach x amount of rewards / fees weekly.
3.   Delegators to have a clearer idea of how much fees they can earn based on their delegated stake and to clearly show the current impact their stake has on the orchestrator.

Also while looking at current orchestrators on the network I realized that the top 4 orchestrators currently hold around 40% of total staked LPT. So if we assume that that there is a direct correlation between staked % and traffic received for an orchestrator, we can conclude that this 4 orchestrators are receiving 40% of all incoming traffic this possibly being more that they can handle. 

Even though it its true that Os can use the -maxSessions flag to prevent being flooded by broadcasters, I think that there is an issue with delegated stake being wasted in this case; if and orchestrator can be at a maximum capacity 100% of the time it is up with X stake, then we can assume that with X + N stake, N stake will we wasted, since N in this case wont help the O get more streams. This case of stake being wasted is more acute when we consider that inflation will slowly tend to 0 over time, becoming transcoding fees the main source of income for Os. The main direct impact of this problem is on delegator rewards, since delegators will see their fee rewards be diluted by a factor of N. 

After further discussion on the discord, Yondon pointed out that it would be useful for the utility to highlight which orchestrators are currently understaked, incentivizing new delegators to stake to understake Os for higher fee rewards. I think we could achieve this by doing the following calculations:

    • Predicted incoming traffic due to stake % vs -maxSessions defined by the O. This could potentially show which Os are undertakes and which are overstaked.
    • % deviation of current eth fees received by O vs expected eth fees. This metric could be interesting since it could showcase that Os are getting more traffic that they should due to other nodes on the network not being able to handle load.

**Describe the solution you are proposing.** 

I have already created small web based application where users can input an orchestrator address and get data for the points 1,2,3 described above. The idea now is to improve on this project by creating the “Theoretical limit page” implementing metrics to showcase under/over staked orchestrators. 

You can find the application here:

https://livepeer-orchestrator-ui.vercel.app/ 

And the code here:

https://github.com/pabloVoorvaart/livepeer-orchestrator-ui

From here on I would love to receive feedback from the livepeer community as to what would be better to implement, but the main idea is to keep the application as is and improve on it.

In the long term, I would also like to add a feature that allows orchestrators to define what hardware they are using so that the app suggests an appropriate -maxSessions value. For this to work, we would need to create some kind of database with supported hardware and their approximate capacity. Orchestrators would then we able to change these values by connecting with their wallet to the app.

**Describe the scope of the project including a rough timeline and milestones** 

In terms of how much the grant should be, I will leave this open for discussion since the current value of LPT has been quite volatile lately. I will most likely use this reward to stake on my own orchestrator that will deployed this week.

For this grant the following will be created:

    • The application is written in JS using Nextjs. The app is currently hosted on Vercel.
    • I am using tailwind for styling.
    • All data for the application is being fetched from thegraph livepeer subraph.
    • All code will be hosted on my github repo.
    • I will provide documentation to facilitate contributions.

I can have it all done 10 days after the grant is accepted.

**Please estimate hours spent on project based on the above** 

Currently I have spent around 10 hours coding the prototype linked above. 

The code needs a mayor refractor due to it not being organized. I will also need to document most of the functions that im using to calculate certain values.

Finally, to add the theoretical limit page I will need another 5 hours to complete this page.

All in all, due to me not being able to commit 100% of my time to the project due to my current work, I estimate that 10 days after the grant is accepted is an acceptable time window to complete this.

**Comment by adamsoffer** [comment]
Hey @pabloVoorvaart - we're really excited about this tool. The idea of showing metrics showcasing which orchestrators are under/over staked is a great one and something the delegator/orchestrator community will love. The committee is currently reviewing the proposal and will get back to you ASAP.

**Comment by chris-remus** [comment]
This looks like a very useful tool and would complement [the business model we built](https://forum.livepeer.org/t/livepeer-orchestrator-business-model-updates-and-grant-progress/1083) nicely.

**Comment by adamsoffer** [comment]
Hey @pabloVoorvaart - thanks again for this thoughtful proposal. We think this will be a very useful tool for new and existing participants of the network, and we're excited to see it get built. The grant node committee would like to fund $1500 denominated in LPT for the 15 hours of work as well as an additional retroactive $500 denominated in LPT for the work you've already put into the tool (the exact amount of LPT will be based on the past 30-day average price due to volatility). @ericxtang mentioned you also built a video player + analytics frontend. As part of this grant, we'd also love to see that get open-sourced and documented. 

Please let us know if you have any questions and if there's anything you need from us to help you get going!

**Comment by pabloVoorvaart** [comment]
@adamsoffer Hey Adam, that is great news!

I will get started with this as soon as possible. There are still some things I need to figure out:

- Adjust eth fee gains formula to account for node performance.
- Is there any ETA regarding the wei / pixel data being available again on the livepeer graphql api?

I will update this issue with the project progress.

Also for the player I can create a new issue and discuss what features you would like to see implemented. Currently we are testing a version that does client side failover for livepeer.com endpoints and collects analytics with google analytics. Its still a work in progress but the end goal is to have it open sourced so that the community can contribute to it.

**Comment by adamsoffer** [comment]
Fantastic! I'll keep you posted on the orchestrator price API status. Should be up very soon.

**Comment by adamsoffer** [comment]
Hey @pabloVoorvaart, sorry for the delay. The orchestrator price rest API is up. The README can be found here: https://github.com/livepeer/orchestrator-price-api#table-of-contents

**Comment by pabloVoorvaart** [comment]
@adamsoffer Cool! Ill get started ASAP then.

**Comment by pabloVoorvaart** [comment]
Hello! I've been very busy lately, but i've managed to get some work done.

This are the changes I have implemented so far:

- Rewrote the whole project to typescript and refactored it to make the code cleaner and more accessible.
- Made ui improvements, such as optimizing the mobile view, and QOL changes for the user experience.
- Started documenting the code by adding a readme and some comments for the most important parts of the codebase.
- Implement the new pricing API into the tool so that we can fetch price per pixel correctly.

Now I will get started with thinking about how to optimize the calculations for the ETH fee earnings  to better match the current behavior of the network. I will also start implementing the calculations for the theoretical limit concept I proposed in the grant proposal above.

Thanks!
